nighthawk/test/statistic_test.cc

```diff
@@ -0,0 +1,209 @@
+    TypeParam a;
+    double offset = std::pow(10, exponential);
+    for (int value : values) {
+      a.addValue(offset + value);
+    }
+    // If an implementation makes this claim, we put it to the test. SimpleStatistic is simple and
+    // fast, but starts failing this test when exponential equals 8. HdrStatistic breaks at 5.
+    // TODO(oschaaf): evaluate ^^
+    if (a.resistsCatastrophicCancellation()) {
+      Helper::expectNear(22.5, a.pvariance(), a.significantDigits());
+      Helper::expectNear(4.7434164902525691, a.pstdev(), a.significantDigits());
+    }
+  }
+}
+
+TYPED_TEST(TypedStatisticTest, OneMillionRandomSamples) {
+  std::mt19937_64 mt(1243);
+  // TODO(oschaaf): Actually the range we want to test is a factor 1000 higher, but
+  // then catastrophical cancellation make SimpleStatistic fail expectations.
+  // For now, we use values that shouldn't trigger the phenomena. Revisit this later.
+  std::uniform_real_distribution<double> dist(1ULL, 1000ULL * 1000 * 60);
+  StreamingStatistic referenceStatistic;
+  TypeParam testStatistic;
+
+  for (int i = 0; i < 999999; ++i) {
+    auto value = dist(mt);
+
+    // Small selftest that we generate a deterministic set in this test.
+    if (i == 10000) {
+      EXPECT_DOUBLE_EQ(13944017.313468568, value);
+    }
+    referenceStatistic.addValue(value);
+    testStatistic.addValue(value);
+  }
+  Helper::expectNear(referenceStatistic.mean(), testStatistic.mean(),
+                     testStatistic.significantDigits());
+  Helper::expectNear(referenceStatistic.pvariance(), testStatistic.pvariance(),
+                     testStatistic.significantDigits());
+  Helper::expectNear(referenceStatistic.pstdev(), testStatistic.pstdev(),
+                     testStatistic.significantDigits());
+}
+
+TYPED_TEST(TypedStatisticTest, ProtoOutput) {
+  TypeParam a;
+
+  a.addValue(6543456);
+  a.addValue(342335);
+
+  const nighthawk::client::Statistic proto = a.toProto();
+
+  EXPECT_EQ(proto.count(), 2);
+  EXPECT_EQ(proto.mean().nanos(), std::round(a.mean()));
+  EXPECT_EQ(proto.pstdev().nanos(), std::round(a.pstdev()));
+}
+
+class StatisticTest : public testing::Test {};
+
+TEST(StatisticTest, HdrStatisticPercentilesProto) {
+  HdrStatistic statistic;
+  int max = 10;
+
+  for (int i = 1; i <= max; i++) {
+    statistic.addValue(i);
+  }
+
+  std::string str;
+  google::protobuf::util::JsonPrintOptions options;
+  google::protobuf::util::MessageToJsonString(statistic.toProto(), &str, options);
+  std::ifstream myfile;
+  // TODO(oschaaf): If this is confirmed to work in Google's environment, wrap
+  // Envoy::TestEnvironment::runfilesPath in a call within the Nighthawk namespace for re-use.
+  // TODO(oschaaf): Set up reusable helpers for .gold file testing, we'll probably get more of that.
+  // Also, move out the test data into its own directory.
+  myfile.open(Envoy::TestEnvironment::runfilesPath("nighthawk/test/hdr_proto_json.gold"));
```

Comment:
Nit: you can also reuse https://github.com/envoyproxy/envoy/blob/master/source/common/filesystem/filesystem_impl.h#L52.